run go test panic

```
go test -v ./...                                                                                                                                                   [22:48:05]
# github.com/ooclab/es/ecrypt
ecrypt/cipher_test.go:13:3: c.encrypt undefined (type *Cipher has no field or method encrypt, but does have Encrypt)
ecrypt/cipher_test.go:14:3: c.decrypt undefined (type *Cipher has no field or method decrypt, but does have Decrypt)
ecrypt/cipher_test.go:26:3: c.encrypt undefined (type *Cipher has no field or method encrypt, but does have Encrypt)
ecrypt/cipher_test.go:27:3: c.decrypt undefined (type *Cipher has no field or method decrypt, but does have Decrypt)
# github.com/ooclab/es/example/tcp/client
example/tcp/client/client.go:44:15: not enough arguments in call to l.OpenTunnel
	have (string, int, string, int, bool)
	want (string, string, int, string, int, bool)
=== RUN   Test_BaseConn
time="2018-04-01T22:48:30+08:00" level=info msg="           3         54.957µs         0.052059 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="           5         11.851µs         0.402360 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="           9          9.521µs         0.901488 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          17         24.624µs         0.658401 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          33         10.655µs         2.953660 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          65           9.46µs         6.552731 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         129          9.476µs        12.982692 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         257          17.31µs        14.159116 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         513         16.817µs        29.091688 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        1025         16.042µs        60.934807 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        2049         16.758µs       116.605721 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        4097          68.94µs        56.675423 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        8193         27.436µs       284.788368 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       16385         48.321µs       323.378110 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       32769         93.932µs       332.697629 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       65537         179.44µs       348.311155 M/s"
--- PASS: Test_BaseConn (0.02s)
=== RUN   Test_SafeConn
time="2018-04-01T22:48:30+08:00" level=info msg="           3         18.025µs         0.158725 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="           5         15.411µs         0.309414 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="           9         10.096µs         0.850145 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          17         20.688µs         0.783665 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          33         16.231µs         1.938960 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="          65         10.415µs         5.951880 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         129         18.904µs         6.507828 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         257         15.695µs        15.616075 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="         513         18.249µs        26.808862 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        1025         14.051µs        69.569153 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        2049          25.52µs        76.570481 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        4097         39.514µs        98.881502 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="        8193         82.913µs        94.236774 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       16385        127.113µs       122.929627 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       32769        218.471µs       143.043945 M/s"
time="2018-04-01T22:48:30+08:00" level=info msg="       65537         445.07µs       140.429491 M/s"
--- PASS: Test_SafeConn (0.02s)
PASS
ok  	github.com/ooclab/es	0.045s
FAIL	github.com/ooclab/es/ecrypt [build failed]
=== RUN   Test_msgSending_IterBufferd
--- PASS: Test_msgSending_IterBufferd (0.17s)
=== RUN   Test_msgSending_GetSegmentByOrderID
--- PASS: Test_msgSending_GetSegmentByOrderID (0.22s)
=== RUN   Test_msgRecving_Save
--- PASS: Test_msgRecving_Save (0.11s)
=== RUN   Test_msgRecving_GetMissing
--- PASS: Test_msgRecving_GetMissing (0.13s)
=== RUN   Test_Socket_RecvMsg_SendMsg
--- PASS: Test_Socket_RecvMsg_SendMsg (3.04s)
=== RUN   Test_Socket_WriterReaderCloser
--- PASS: Test_Socket_WriterReaderCloser (12.55s)
PASS
ok  	github.com/ooclab/es/proto/udp	16.282s
link quit:  <nil>
link quit:  <nil>
time="2018-04-01T22:48:30+08:00" level=error msg="Link.recv quit" error=EOF from=link id=0
time="2018-04-01T22:48:30+08:00" level=error msg="Link.send quit" error="get nil from l.outbound" from=link id=0
time="2018-04-01T22:48:30+08:00" level=error msg="send request to remote endpoint failed" error="unexpected end of JSON input"
time="2018-04-01T22:48:30+08:00" level=error msg="OpenTunnel failed: unexpected end of JSON input"
panic: send on closed channel

goroutine 20 [running]:
github.com/ooclab/es/session.(*Manager).HandleIn(0xc42013c040, 0xc4201a4001, 0xcb, 0xcb, 0x0, 0x0)
	/Users/lowstz/code/golang/src/github.com/ooclab/es/session/manager.go:38 +0x386
github.com/ooclab/es/link.(*Link).recv(0xc420142000, 0x137d720, 0xc42014a090, 0x0, 0x0)
	/Users/lowstz/code/golang/src/github.com/ooclab/es/link/link.go:279 +0x40a
github.com/ooclab/es/link.(*Link).Bind.func1(0xc420142000, 0x137d720, 0xc42014a090)
	/Users/lowstz/code/golang/src/github.com/ooclab/es/link/link.go:329 +0x5e
created by github.com/ooclab/es/link.(*Link).Bind
	/Users/lowstz/code/golang/src/github.com/ooclab/es/link/link.go:327 +0xbd
FAIL	github.com/ooclab/es/test	0.020s
```

--- a/ecrypt/cipher_test.go
+++ b/ecrypt/cipher_test.go
@@ -10,8 +10,8 @@ func TestRC4(t *testing.T) {
 	c := NewCipher("rc4", secret)
 	dst := make([]byte, len(clearText))
 	dst2 := make([]byte, len(clearText))
-	c.encrypt(dst, []byte(clearText))
-	c.decrypt(dst2, dst)
+	c.Encrypt(dst, []byte(clearText))
+	c.Decrypt(dst2, dst)
 	if clearText != string(dst2) {
 		t.Error(string(dst2))
 	}
@@ -23,8 +23,8 @@ func TestAES256CFB(t *testing.T) {
 	c := NewCipher("aes256cfb", secret)
 	dst := make([]byte, len(clearText))
 	dst2 := make([]byte, len(clearText))
-	c.encrypt(dst, []byte(clearText))
-	c.decrypt(dst2, dst)
+	c.Encrypt(dst, []byte(clearText))
+	c.Decrypt(dst2, dst)
 	if clearText != string(dst2) {
 		t.Error(string(dst2))
 	}

--- a/example/tcp/client/client.go
+++ b/example/tcp/client/client.go
@@ -37,11 +37,11 @@ func main() {
 		// remoteHost := "127.0.0.1"
 		// remotePort := 3000
 		// reverse := false
-		localHost, localPort, remoteHost, remotePort, reverse, err := parseTunnel(os.Args[2])
+		proto, localHost, localPort, remoteHost, remotePort, reverse, err := parseTunnel(os.Args[2])
 		if err != nil {
 			panic(err)
 		}
-		l.OpenTunnel(localHost, localPort, remoteHost, remotePort, reverse)
+		l.OpenTunnel(proto, localHost, localPort, remoteHost, remotePort, reverse)
 	}()
 
 	ec := es.NewBaseConn(conn)
@@ -49,7 +49,7 @@ func main() {
 	l.Close()
 }
 
-func parseTunnel(value string) (localHost string, localPort int, remoteHost string, remotePort int, reverse bool, err error) {
+func parseTunnel(value string) (proto string, localHost string, localPort int, remoteHost string, remotePort int, reverse bool, err error) {
 	L := strings.Split(value, ":")
 	if len(L) != 5 {
 		fmt.Println("tunnel format: \"r|f:local_host:local_port:remote_host:remote_port\"")
@@ -66,6 +66,15 @@ func parseTunnel(value string) (localHost string, localPort int, remoteHost stri
 		reverse = false
 	default:
 		err = errors.New("wrong tunnel map")
+		return
+	}
+
+	proto = strings.TrimSpace(strings.ToLower(L[1]))
+	if proto == "" {
+		proto = "tcp"
+	}
+	if !(proto == "tcp" || proto == "udp") {
+		err = errors.New("unknown protocol")
 		return
 	}
 	localPort, err = strconv.Atoi(L[2])

--- a/test/common.go
+++ b/test/common.go
@@ -46,9 +46,13 @@ func getServerAndClient() (serverLink *link.Link, clientLink *link.Link, err err
 				l := link.NewLink(nil)
 				serverLinkCh <- l
 				ec := es.NewBaseConn(conn)
-				l.Bind(ec)
-				l.Close()
-				fmt.Println("link quit: ", err)
+				if err := l.Bind(ec); err != nil {
+					logrus.Errorf("link bind failed: %s", err)
+				}
+				l.Wait()
+				if err := l.Close(); err != nil {
+					logrus.Errorf("link quit: ", err)
+				}
 			}()
 		}
 	}()
@@ -86,8 +90,10 @@ func connectServer(addr string) *link.Link {
 	ec := es.NewBaseConn(conn)
 	// FIXME: quit it not a good choice for testcase!
 	go func() {
-		l.Bind(ec)
-		fmt.Println("link quit: ", err)
+		if err := l.Bind(ec); err != nil {
+			logrus.Errorf("link quit: ", err)
+		}
+		l.Wait()
 		l.Close()
 	}()
 	return l

--- a/test/session_test.go
+++ b/test/session_test.go
@@ -53,7 +53,10 @@ func Benchmark_LinkInnerSessionMulti(b *testing.B) {
 }
 
 func Test_LinkInnerSessionMinimalFrame(t *testing.T) {
-	_, clientLink, _ := getServerAndClient()
+	_, clientLink, err := getServerAndClient()
+	if err != nil {
+		t.Error(err)
+	}
 
 	s, _ := clientLink.NewSession()
 	for i := 0; i < 1025; i++ {

